integration/test_web.py

```diff
@@ -0,0 +1,496 @@
+
+    resp = requests.post(
+        dircap_uri,
+        data={
+            u"t": u"upload",
+            u"when_done": u".",
+        },
+        files={
+            u"file": FILE_CONTENTS,
+        },
+    )
+
+    # confirm the file is in the directory
+    resp = requests.get(
+        dircap_uri,
+        params={
+            u"t": u"json",
+        },
+    )
+    d = json.loads(resp.content)
+    k, data = d
+    assert k == u"dirnode"
+    assert len(data['children']) == 1
+    k, child = data['children'].values()[0]
+    assert k == u"filenode"
+    assert child['size'] == len(FILE_CONTENTS)
+
+    # perform deep-stats on it...
+    resp = requests.post(
+        dircap_uri,
+        data={
+            u"t": u"start-deep-stats",
+            u"ophandle": u"something_random",
+        },
+    )
+    assert resp.status_code >= 200 and resp.status_code < 300
+
+    # confirm we get information from the op .. after its done
+    while True:
+        resp = requests.get(
+            util.node_url(alice.node_dir, u"operations/something_random"),
+        )
+        d = json.loads(resp.content)
+        if d['size-literal-files'] == len(FILE_CONTENTS):
+            print("stats completed successfully")
+            break
+        else:
+            print("{} != {}; waiting".format(d['size-literal-files'], len(FILE_CONTENTS)))
+        time.sleep(.5)
+
+
+def test_status(alice):
+    """
+    confirm we get something sensible from /status and the various sub-types
+    """
+
+    # upload a file
+    # (because of the nature of the integration-tests, we can only
+    # assert things about "our" file because we don't know what other
+    # operations may have happened in the grid before our test runs).
+
+    FILE_CONTENTS = u"all the Important Data of alice\n" * 1200
+
+    resp = requests.put(
+        util.node_url(alice.node_dir, u"uri"),
+        data=FILE_CONTENTS,
+    )
+    cap = resp.text.strip()
+
+    print("Uploaded data, cap={}".format(cap))
+    resp = requests.get(
+        util.node_url(alice.node_dir, u"uri/{}".format(urllib2.quote(cap))),
+    )
+
+    print("Downloaded {} bytes of data".format(len(resp.content)))
+    assert resp.content == FILE_CONTENTS
+
+    resp = requests.get(
+        util.node_url(alice.node_dir, "status"),
+    )
+    dom = html5lib.parse(resp.content)
+
+    hrefs = [
+        a.get('href')
+        for a in dom.iter(u'{http://www.w3.org/1999/xhtml}a')
+    ]
+
+    found_upload = False
+    found_download = False
+    for href in hrefs:
+        if href.startswith(u"/") or not href:
+            continue
+        resp = requests.get(
+            util.node_url(alice.node_dir, u"status/{}".format(href)),
+        )
+        if href.startswith(u'up'):
+            assert "File Upload Status" in resp.content
+            if "Total Size: {}".format(len(FILE_CONTENTS)) in resp.content:
+                found_upload = True
+        elif href.startswith(u'down'):
+            print(href)
```

Comment:
What's the story behind prints in the integration tests?  A good way to gather helpful debug information?  Or a bad way that leaves a mess?

Reply:
Well, that one is spurious and I've deleted it.

But in general, `print()` is I think a fine way to output useful / debug information since `py.test` captures it per-test and you only see it if you want to (i.e. via `py.test -s`). Most of the logs are output this way (and some different logs are captured by the `flogtool gather` process that a fixture sets up).

e.g. compare `py.test -v integration` output to `py.test -sv integration` output.